ui/components/src/JSONEditor.tsx

```diff
@@ -12,45 +12,42 @@
 // limitations under the License.
 
 import { useEffect, useState } from 'react';
-import { TextField } from '@mui/material';
+import CodeMirror from '@uiw/react-codemirror';
+import { json, jsonParseLinter } from '@codemirror/lang-json';
+import { linter, lintGutter } from '@codemirror/lint';
+import { useTheme } from '@mui/material';
 
 interface JSONEditorProps<Spec> {
   value: Spec;
   onChange?: (next: Spec) => void;
 }
 
 export function JSONEditor<T>(props: JSONEditorProps<T>) {
+  const theme = useTheme();
+  const isDarkMode = theme.palette.mode === 'dark';
+
   const [value, setValue] = useState(() => JSON.stringify(props.value, null, 2));
-  const [invalidJSON, setInvalidJSON] = useState(false);
 
   useEffect(() => {
     setValue(JSON.stringify(props.value, null, 2));
-    setInvalidJSON(false);
   }, [props.value]);
 
-  // TODO: replace with CodeMirror editor
   return (
-    <TextField
-      label="JSON"
-      error={invalidJSON}
-      helperText={invalidJSON ? 'Invalid JSON' : ''}
-      multiline
-      fullWidth
+    <CodeMirror
+      theme={isDarkMode ? 'dark' : 'light'}
+      extensions={[json(), linter(jsonParseLinter()), lintGutter()]}
       value={value}
-      onChange={(event) => {
-        setValue(event.target.value);
+      onChange={(newValue) => {
+        setValue(newValue);
       }}
-      maxRows={20}
       onBlur={() => {
         try {
           const json = JSON.parse(value ?? '{}');
-          setInvalidJSON(false);
           if (props.onChange !== undefined) {
             props.onChange(json);
           }
-        } catch (e) {
-          setInvalidJSON(true);
-        }
+          // eslint-disable-next-line no-empty
+        } catch (e) {}
```

Comment:
Do we need to do something if this throws to communicate the JSON is bad or is that all handled internally in codemirror? If the latter, might be worth putting a comment inside the catch that clarifies that. As a fun side effect, I think putting a comment inside the catch is enough to not need the `eslint-disable`.

Reply:
CodeMirror's json extension handles the JSON linting/communication by showing where the failure happens directly in the editor. 

> As a fun side effect, I think putting a comment inside the catch is enough to not need the eslint-disable.

Oh nice, I can do that too.